src/fs.rs

```diff
@@ -0,0 +1,449 @@
+        }
+    }
+
+    /// Renames a file from `old_name` to `new_name`.
+    fn rename(old_name: &wstr, new_name: &wstr) -> std::io::Result<()> {
+        if wrename(old_name, new_name) == -1 {
+            let error_number = errno::errno();
+            FLOG!(
+                error,
+                wgettext_fmt!("Error when renaming file: %s", error_number.to_string())
+            );
+            return Err(std::io::Error::from(error_number));
+        }
+        Ok(())
+    }
+
+    const TMP_FILE_SUFFIX: &wstr = L!(".XXXXXX");
+    let mut tmp_file_template = path.to_owned();
+    tmp_file_template.push_utfstr(TMP_FILE_SUFFIX);
+    let tmp_file_template = tmp_file_template;
+    let (tmp_file, tmp_name) = create_temporary_file(&tmp_file_template)?;
+    // Ensure that the temporary file is unlinked when this function returns.
+    let mut tmp_file = ScopeGuard::new(tmp_file, |_| {
+        wunlink(&tmp_name);
+    });
+
+    // We want to rewrite the file.
+    // To avoid issues with crashes during writing,
+    // we write to a temporary file and once we are done, this file is renamed such that it
+    // replaces the original file.
+    if locking_enabled.load() {
+        // To avoid races, we need to have exclusive access to the file for the entire
+        // duration, which we get via an exclusive lock on the corresponding lock file.
+        // Taking a shared lock first and later upgrading to an exclusive one could result in a
+        // deadlock, so we take an exclusive one immediately.
+        match LockedFile::new(
+            LockingMode::Exclusive,
+            path,
+            OFlag::O_RDONLY | OFlag::O_CREAT,
+            Mode::S_IRUSR | Mode::S_IWUSR,
+        ) {
+            Ok(mut locked_file) => match rewrite(locked_file.get_mut(), &mut tmp_file) {
+                Ok(update_data) => {
+                    update_metadata(locked_file.get(), &tmp_file);
+                    rename(&tmp_name, path)?;
+                    return Ok((file_id_for_path(path), update_data));
+                }
+                Err(e) => {
+                    return Err(e);
+                }
+            },
+            Err(e) => {
+                FLOGF!(
+                    synced_file_access,
+                    "Error trying to lock file '%s': %s",
+                    path,
+                    e,
+                );
+                locking_enabled.store(false);
+            }
+        }
+    }
+
+    // If this is reached, we assume that locking is not available so we use a fallback
+    // implementation which tries to avoid race conditions, but in the case of contention it is
+    // possible that some writes are lost.
+
+    FLOG!(
+        synced_file_access,
+        "flock-based locking is disabled. Using fallback implementation."
+    );
+
+    // Give up after this many unsuccessful attempts.
+    // TODO: which value, should this be a constant shared with other retrying logic?
+    let max_attempts = 1000;
```

Comment:
regarding the TODO, do we still have similar retrying logic elsewhere? Can't think of any that doesn't retry indefinitely

Reply:
right. In a way they're doing the same thing, so yeah I'd think extracting makes sense.

Actually, it might help to extract a function for the retry loop, checking file IDs etc.
Not a blocker, but if we ever need to touch this logic again, we might want to try extracting that function first